### DISCUSSION TITLE
MPS20N0040D - Lib in micropython to use this sensor.

### DISCUSSION THREAD

**Discussion by pierregraciano** [micropython/micropython#14028] [Hardware & Peripherals] [answered]
Hello everyone.
I bought a MPS20N0040D sensor, and Im searching ways to test it on my ESP32, using micropython. Do you have some website or code to help me to test this sensor? The module has HX710b.
Thank you.

**Comment by robert-hh** [comment]
The MPS20N0040D sensor itself is just a passive resistor bridge. It requires additional circuitry for connecting it to a ESP32. There are breakout boards available which use this sensor. Do you use such a board? Then the type of this board and the interface chip is important.

**Reply by pierregraciano** [reply]
Good morning, Robert. Yeah, Im using a module, like this one (link below). The chip is HX710b.

Link: (https://www.amazon.com/ACEIRMC-Digital-Barometric-Pressure-Controller/dp/B09SYQMJD8/ref=sr_1_2?crid=3N58R2Z9UUIHT&dib=eyJ2IjoiMSJ9.eXJPsg0_QVc7OWt_bmNcTDb418pnWGqlJ652V7hF1tFmsojhjSnF_rJZc91YXrun.nETXynehj2i4HWsc2PtvccOOGBMy5zZIcdd7dSNgAWs&dib_tag=se&keywords=MPS20N0040D&qid=1709641155&sprefix=mps20n0040d%2Caps%2C234&sr=8-2)

**Reply by pierregraciano** [reply]
I found some lib in C++, but not in MicroPython.

**Reply by robert-hh** [reply]
The chip interface looks identical to the HX711, for which drivers exist. Only the relation between clock pulses and result must be interpreted differently. On the HX711, it controls input and gain, on the HX710 it controls the source.
But generally, a HX711 driver should work.

**Reply by pierregraciano** [reply]
Sorry for the inconvenience but could you send me a sample code?

**Reply by robert-hh** [reply]
There are a few, but I point at the one I have adapted: https://github.com/robert-hh/hx711

You can use the hx711_gpio.py variant and follow the example in README.md for micropython.org, gpio mode. You can use any pin pair for that. With the `gain=xxx` you can control the mode. With `gain=128` (default) or `gain=64` you get the pressure sensor reading, with gain=32 you get the temperature reading (hx710A) or VDDIO-VDDA reading (hx710b).
With read() or read_average() you get the raw sensor reading. Calibration (hPa per reading) must be done outside driver.
The results will look noisy. Even if the HX710 reading is 24bit, only the higher 16 - 17 bits are reliable.

**Reply by robert-hh** [reply]
I just uploaded a variant for the HX710 to https://github.com/robert-hh/hx710. It is untested! Instead of gain=xxx you have to use mode=n, with mode=1 for the external sensor as default.

**Reply by robert-hh** [reply]
Tested now with a HX711. The driver runs. Whether the results are reasonable I cannot tell. I ordered a breakout board like yours, but it will take 2-3 weeks before it arrives.

**Reply by pierregraciano** [reply]
Thank you so much, Robert. I'll try this now and come here later to bring you news. Thanks for everything!!!!

**Reply by pierregraciano** [reply]
The code is running but I have one question: Which function I use to get the real pressure in KPa (kilopascal)? 
Im using `value2 = hx710.get_value()` and the console shows:
```
110.25
62.5
104.25
86.5
17.25
78.75
4.75
3.75
```
Maybe a noise in system?

**Reply by robert-hh** [reply]
You should use hx710.read() or hx710.read_average(). To get the real pressure, you have to determine scaling value by comparing the read value with an actual pressure from a local reference measurement. Take care that you get as reference the absolute pressure values. In the usual weather report the sealevel pressure is reported.

The values you got above look like noise.

**Reply by pierregraciano** [reply]
Ok, thank you. Later I will try with a air pump and see the results.

**Reply by robert-hh** [reply]
Reading the data sheet of the MPS20N0040D again I noticed, that it's a differential sensor. So your reading above are just noise at a expected level (7 bits of the 24) when reading a zero pressure difference.

**Reply by pierregraciano** [reply]
Hi Robert. When I turn on my sensor, these numbers appears:
```
-4037159
-4037145
-4037458
```
»»» log2 4037458 ≈ 22


This is the environmental pressure maybe. When I blow air into the sensor, the numbers are:
```
8388607
8388607
8388607
```
»»» log2 8388607 ≈ 23

Im using the code below. Its possible to calibrate my sensor? If these numbers are ADC values, how can I convert to KPa? 
```
from hx710 import HX710
from machine import Pin

pin_OUT = Pin(12, Pin.IN, pull=Pin.PULL_DOWN)
pin_SCK = Pin(13, Pin.OUT)

hx710 = HX710(pin_SCK, pin_OUT)

while True:
  hx710.tare()
  value = hx710.read()
  print(value)
```

**Reply by robert-hh** [reply]
Since the MPS20N0040D is a differential sensor, the reading without a pressure applied is just the base reading. I'm surprised that it is -4037159, close to the limits of the measuring range. I would have expected a value of about 0. Please be aware that 8388607 is equivalent to 0x7fffff. So this means overflow for a signed 24bit integer.

To get a hPa value out of the readings, you have to apply a linear transformation:

`hpa = (value - tare) / scale`

Where  `tare` is the reading without an applied pressure, and `scale` is  `delta-reading/delta_pressure`. The delta_reading values come from the HX710, like in your example `8388607 - -4037159`, and for delta presure you need a know pressure.
The MPS20N0040D is just a resistor bridge, the HX710 is just a differential amplifier plus ADC. Both do not know of the physical quantities that cause their output or input to vary. You are the one that has to know the relation.

**Reply by robert-hh** [reply]
Reading the data sheet, the pressure range is 40kPa or 400hPa. That would exclude applications named in the data sheet like tire pressure, car air pump, which have typical pressures of 2000 - 4000 hPa.
The offset at the sensor may be +/+25mV. At an amplification of 128 for the HX128 this is 3.2V. So yes, the offset can be close to the full range of the HX710. 
I will get my copies of the breakout board in a week or two. But looking at the data, I expect it to be just food for the trash can.
I worked with a very good pressure sensor from DLV, the DS-0036. Very reliable, calibrated, with various models depending on the pressure range. I have a repository for that sensor here: https://github.com/robert-hh/DLV-pressure-sensor You'll find a data sheet there as well.
The sensor is expensive but provides reliable results.

**Reply by robert-hh** [reply]
For calibration you may be interested in this link, which you might have seen already: https://makersportal.com/blog/2020/6/4/mps20n0040d-pressure-sensor-calibration-with-arduino
The Arduino just serves as means to read data from the HX710. You can use your MicroPython module instead.

**Reply by robert-hh** [reply]
So I got the modules today. A quick first test showed:
- the transfer ratio reading to pressure is about 67 per Pa, or 67000 per kPa
- the quiet reading is about -1816320 at an input voltage of  -2.7mV
- the ratio mV/Pa is about 1mV/1kPa.
- the hx710 is limited to about +/- 12mV at the input. The sensor showed ~100mV at 100kPa, and about 40mV at 40kPa. So using the hx710 is a design flaw. The hx711 with a gain of 32 would be much better, giving a +/-48mV range, or even a ADS1114 or ADS1115 in differential +/-256mV mode. That would at least result in 14 stable bits.
- The module did not use the option of the sensor to implement a zero offset trim. For that purpose at the sensor pins 1 and 6 as part of the bridge a separately fed out. You could use a ~2k trim potentiometer to adjust the 0 pressure offset.
 
Summary: The sensor seems OK, the module is a poor design. 

Note: The pressures are a rough estimation. At first, I just use a Bicycle pump with a scale.

**Reply by pierregraciano** [reply]
Thank you again, Robert. Im testing with Arduino now and after this I will try to code in micropython. For this test did you use your own hx710 lib?

**Reply by robert-hh** [reply]
Yes. I used my own lib and read_average().
Edit: Is the range of about +/-10kPa suitable for you application?

**Reply by pierregraciano** [reply]
Ok, I'll try it using a manometer.

**Reply by pierregraciano** [reply]
I think the reading needs to be ample, why?

**Reply by robert-hh** [reply]
The pressure range for which the module can be used is +/- 10kPa, about 1/10 of the atmospheric pressure. That has to fit to your application.

**Reply by pierregraciano** [reply]
What limits this pressure range is the HX710B?

**Reply by robert-hh** [reply]
Yes. The input range of the HX710 is about +/-12mV  (1.65V / 128) at 3.3V supply. At 5V supply it's about +/- 18mV, but at the end of the range it get's nonlinear, so you cannot use the full range.

**Reply by pierregraciano** [reply]
Do you think is a good idea buy some HX711 and use in place of HX710? The footprints are different, but some wires can fix it.

**Reply by robert-hh** [reply]
If you need the larger range, then you have to change your approach. You can get as well just the MPS20N040 sensor itself in a DIP package, 2.54mm pitch, and use these with a HX711 breakout at a gain of 32 or another sensitive A/D converter breakout like the ADS1115, which has +/-256,V as lowest input range. You can test that easily on a breakout board.

**Reply by robert-hh** [reply]
There is another quick hack: if you have good eyes and a steady hand, you can solder a resistor across the sensor output. The source impedance of the sensor bridge is about 5k, so a 1.5k resistor should increase the range by about 4. That will of course have a little bit more noise than a HX711 at gain 32, but less than the other noise sources.
At the board here it's between Pins 4 and 6 of the sensor.

**Comment by robert-hh** [comment]
I made two test series comparing the MPS20N040D sensor and a calibrated DLV0105D sensor attached to the same pressure volume. The volume has a small leak, so the pressure falls within a suitable time, giving a nice range test. I used two variants of the MPS20N04D sensor: a) raw sensor attached to a HX710B, b) Sensitivity reduced sensor by attaching a 1.5k resistor across pins 4 and 6 attached to a HX710B.

a) Raw sensor: The resulting conversion factor of MPS20N04D at HX710B reading to kPa was 710000. The pressure range is about +/- 12 kPa. A second MSP20N040D sensor had a conversion factor of 710000.
b) Sensitivity reduced sensor. The resulting conversion factor of MPS20N04D at HX710B reading to kPa was 174380. The pressure range is about +/- 48 kPa.

In the test, the MPS20N040D was sampled first, then the DLV015D. Some diagrams of the test series below. The diagrams show, that both sensors run pretty synchronous. 

a) Raw sensor:

![grafik](https://github.com/micropython/micropython/assets/12476868/b394cfa9-7b8b-4f3f-aceb-ef2bfd267453)

Detail at the low pressure end of the test:

![grafik](https://github.com/micropython/micropython/assets/12476868/1b2823e6-87ac-4bbb-a95d-de35bb700fac)

b) Wider range sensor using a 1k5 resistor across the sensor output.

![grafik](https://github.com/micropython/micropython/assets/12476868/bb2330db-302d-4b7d-a1fc-336e5ebe1e96)

Detail at the low pressure end of the test:

![grafik](https://github.com/micropython/micropython/assets/12476868/c5af04c4-4811-496d-985f-6bcbb6a627d7)

**Comment by robert-hh** [comment]
By the way: the data sheet of the MPS20N040D is inconsistent. It mentions 540PSI and 40kPa, but 40kPa is equivalent to about 5.4 PSI, or the other way around: 540PSI is equivalent to about 4000 kPa or 40bar.

**Reply by robert-hh** [reply]
As another option to extend the range of the MPS20N0040D, the voltage across the bridge can be reduced, lowering the output range. A test of setting the voltage to 1.25V using a voltage reference results in a pressure range of +/-24kPa. The test results are good. Still, lowering the output with a resistor seems the easier way, and as shown in the diagrams above, the results are fine.

**Reply by robert-hh** [reply, marked as answer]
Some figures of the sensor with the voltage at 1.25V.

Full test series:

![grafik](https://github.com/micropython/micropython/assets/12476868/d19178b1-8f74-419b-afc8-54d2f0d4c3bc)

Detail at the lower range:

![grafik](https://github.com/micropython/micropython/assets/12476868/c1427ebd-f5da-4a9e-9159-f6f568896953)

**Comment by robert-hh** [comment]
I got samples of the MPS20N0040D sensor and wired them up to a) and ADS1115 A/D converter in the +/-256mV range, and b) an HX711 at the B inputs with a gain of 32. Both variants cover the full range of the pressure sensor, and the results show good tracking with the DLV-015D sensor, as above, with higher resolution than the DLV-015D.

Test series with HX711:
![grafik](https://github.com/micropython/micropython/assets/12476868/1f3099c0-dfc0-4a40-82af-5d3f923f9096)

Detail at low pressure:
![grafik](https://github.com/micropython/micropython/assets/12476868/e004ab2c-9527-4b48-ae0b-50cb03930236)

**Reply by pierregraciano** [reply]
Hello Robert. Thank you so much, you will help me a lot! I'll mark your answer as correct. Thank you again!!

**Comment by robert-hh** [comment]
Maenwhile I made another test using the raw MPS20N004D sensor connected to a ADS1115 ADC at a pressure up to 100kPa. Diagram below, as always compared to the DLB015d. Since the pressure falls almost in a e**-t/T series, a log/lin scaling for the results should be close to a straight line. The output of the mps20N0040D is shown as mV. The yellow line the is ratio between the output of MPS20N0040D and the DLV015D. It should be mostly constant. Since the DLV015D has a lower resolution, about 10.5 digits, the results at low pressure get worse.

![grafik](https://github.com/micropython/micropython/assets/12476868/ead01d44-a531-4645-a781-7692e6e0f808)
